File: packages/python/src/kimpton_evalrouter/_transport.py
"""Bounded requests to one explicitly configured API origin."""

from __future__ import annotations

import ipaddress
import json
import math
import os
import random
import re
import threading
import time
from contextlib import suppress
from dataclasses import dataclass
from datetime import UTC, datetime
from email.utils import parsedate_to_datetime
from urllib.parse import quote, urlsplit, urlunsplit

import httpx


class ClientError(Exception):
    idempotency_key: str | None = None

    def __init__(self, code: str, message: str):
        self.code = code
        self.message = message
        super().__init__(message)

    def as_dict(self):
        return {
            "code": self.code,
            "message": self.message,
            **(
                {"idempotency_key": self.idempotency_key}
                if self.idempotency_key is not None
                else {}
            ),
        }


class APIError(ClientError):
    def __init__(
        self,
        status: int,
        code: str,
        request_id: str | None = None,
        *,
        message: str | None = None,
        details=None,
    ):
        self.status = status
        self.request_id = request_id
        self.details = details
        super().__init__(code, message or f"API request failed (HTTP {status}, {code}).")

    def as_dict(self):
        return {
            **super().as_dict(),
            "status": self.status,
            "request_id": self.request_id,
            "details": self.details,
        }


class RequestTimeout(ClientError):
    def __init__(self):
        super().__init__(
            "request_timeout",
            "The request deadline expired; a submitted operation may have committed.",
        )


class WaitCancelled(ClientError):
    def __init__(self):
        super().__init__(
            "wait_cancelled",
            "Local waiting was cancelled. The server evaluation was not cancelled.",
        )


@dataclass(frozen=True)
class RequestOptions:
    timeout: float | None = None
    cancel_event: threading.Event | None = None


def positive(value, label):
    if (
        isinstance(value, bool)
        or not isinstance(value, (int, float))
        or not math.isfinite(value)
        or value <= 0
    ):
        raise ClientError("invalid_configuration", f"{label} must be a positive finite number.")
    return float(value)


def segment(value: str):
    if (
        not isinstance(value, str)
        or not re.fullmatch(r"[A-Za-z0-9_.:@+-]{1,200}", value)
        or value in {".", ".."}
    ):
        raise ClientError(
            "invalid_input", "Resource identifiers must be nonempty path-safe identifiers."
        )
    return quote(value, safe="")


def normalize_base_url(value, allow_http):
    try:
        parts = urlsplit(value)
        host = parts.hostname
        port = parts.port
        if (
            not host
            or parts.username is not None
            or parts.password is not None
            or parts.query
            or parts.fragment
        ):
            raise ValueError()
        if (
            parts.path.rstrip("/") not in {"", "/v1"}
            or "\\" in value
            or any(ord(c) <= 32 for c in value)
        ):
            raise ValueError()
        local = host == "localhost"
        with suppress(ValueError):
            local = local or ipaddress.ip_address(host).is_loopback
        if parts.scheme != "https" and not (parts.scheme == "http" and allow_http and local):
            raise ValueError()
        if port is not None and not 1 <= port <= 65535:
            raise ValueError()
        return urlunsplit((parts.scheme, parts.netloc, "/v1", "", ""))
    except (TypeError, ValueError):
        raise ClientError(
            "invalid_configuration",
            "Set an HTTPS API origin or /v1 URL. Explicit HTTP is allowed only for loopback development.",
        ) from None


def checkpoint(deadline, cancel_event):
    if cancel_event is not None and cancel_event.is_set():
        raise WaitCancelled()
    remaining = deadline - time.monotonic()
    if remaining <= 0:
        raise RequestTimeout()
    return remaining


def pause(delay, deadline, cancel_event):
    remaining = checkpoint(deadline, cancel_event)
    seconds = min(delay, remaining)
    if cancel_event is not None:
        if cancel_event.wait(seconds):
            raise WaitCancelled()
    else:
        time.sleep(seconds)
    checkpoint(deadline, cancel_event)


def retry_delay(attempt, header):
    delay = min(5, 0.25 * 2**attempt) + random.random() * 0.25
    if header and len(header) < 100:
        try:
            suggested = float(header)
        except ValueError:
            try:
                suggested = (parsedate_to_datetime(header) - datetime.now(UTC)).total_seconds()
            except (ValueError, TypeError, OverflowError):
                suggested = 0
        if math.isfinite(suggested):
            delay = max(delay, suggested)
    return delay


def secret_values(body):
    values = []
    if isinstance(body, dict):
        for key, value in body.items():
            if (
                isinstance(key, str)
                and re.search(r"key|secret|token|password", key, re.I)
                and isinstance(value, str)
                and value
            ):
                values.append(value)
            values.extend(secret_values(value))
    elif isinstance(body, list):
        for value in body:
            values.extend(secret_values(value))
    return values


def safe_error_value(value, secrets, depth=0):
    if depth > 4:
        return "[truncated]"
    if isinstance(value, str):
        for secret in secrets:
            value = value.replace(secret, "[redacted]")
        return "".join(c if c.isprintable() else " " for c in value)[:2000]
    if isinstance(value, dict):
        return {
            safe_error_value(str(key), secrets)[:100]: "[redacted]"
            if re.search(r"key|secret|token|password|authorization|cookie", str(key), re.I)
            else safe_error_value(item, secrets, depth + 1)
            for key, item in list(value.items())[:100]
        }
    if isinstance(value, list):
        return [safe_error_value(item, secrets, depth + 1) for item in value[:100]]
    if isinstance(value, float) and not math.isfinite(value):
        return None
    return value if value is None or type(value) in {int, float, bool} else None


def reject_constant(_value):
    raise ValueError("Non-finite JSON number")


class Transport:
    def __init__(
        self,
        *,
        base_url: str | None = None,
        api_key: str | None = None,
        workspace_id: str | None = None,
        timeout: float = 120,
        request_timeout: float = 30,
        max_retries: int = 2,
        max_response_bytes: int = 64 * 1024 * 1024,
        max_export_bytes: int = 256 * 1024 * 1024,
        allow_http: bool = False,
        transport: httpx.BaseTransport | None = None,
    ):
        self._base = normalize_base_url(
            base_url or os.environ.get("EVALROUTER_BASE_URL", ""), allow_http
        )
        self._key = api_key if api_key is not None else os.environ.get("EVALROUTER_API_KEY")
        self._workspace = (
            workspace_id if workspace_id is not None else os.environ.get("EVALROUTER_WORKSPACE_ID")
        )
        for label, value in (("API key", self._key), ("Workspace ID", self._workspace)):
            if value is not None and (
                not isinstance(value, str)
                or not 1 <= len(value) <= 4096
                or any(not 33 <= ord(c) <= 126 for c in value)
            ):
                raise ClientError(
                    "invalid_configuration", f"{label} must be printable non-space ASCII."
                )
        self.timeout = positive(timeout, "Timeout")
        self.request_timeout = positive(request_timeout, "Request timeout")
        if type(max_retries) is not int or not 0 <= max_retries <= 5:
            raise ClientError("invalid_configuration", "max_retries must be between 0 and 5.")
        self.max_retries = max_retries
        for value in (max_response_bytes, max_export_bytes):
            if type(value) is not int or value < 1:
                raise ClientError(
                    "invalid_configuration", "Response size limits must be positive integers."
                )
        self.max_response_bytes = max_response_bytes
        self.max_export_bytes = max_export_bytes
        self._http = httpx.Client(follow_redirects=False, trust_env=False, transport=transport)

    def close(self):
        self._http.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def request(
        self,
        method,
        path,
        *,
        body=None,
        query=None,
        retry=False,
        idempotency_key=None,
        binary=False,
        options=None,
    ):
        options = options or RequestOptions()
        deadline = time.monotonic() + positive(
            options.timeout if options.timeout is not None else self.timeout, "Timeout"
        )
        if not path.startswith("/") or path.startswith("//"):
            raise ClientError("invalid_input", "Invalid API resource path.")
        headers = {
            "Accept": "*/*" if binary else "application/json",
            "User-Agent": "kimpton-evalrouter-python/0.1.0",
        }
        if self._key:
            headers["Authorization"] = f"Bearer {self._key}"
        if self._workspace:
            headers["X-Workspace-Id"] = self._workspace
        if idempotency_key is not None:
            if (
                not isinstance(idempotency_key, str)
                or not 1 <= len(idempotency_key) <= 128
                or any(not 32 <= ord(c) <= 126 for c in idempotency_key)
            ):
                raise ClientError(
                    "invalid_input",
                    "Idempotency keys must contain 1–128 printable ASCII characters.",
                )
            headers["Idempotency-Key"] = idempotency_key
        try:
            content = (
                None
                if body is None
                else json.dumps(
                    body, ensure_ascii=False, allow_nan=False, separators=(",", ":")
                ).encode()
            )
            parameters = {
                key: str(value).lower() if isinstance(value, bool) else value
                for key, value in (query or {}).items()
                if value is not None
            }
        except (TypeError, ValueError, AttributeError):
            raise ClientError("invalid_input", "Request data must be JSON-serializable.") from None
        if content is not None:
            headers["Content-Type"] = "application/json"
        secrets = [value for value in [self._key, *secret_values(body)] if value]
        for attempt in range(self.max_retries + 1):
            remaining = checkpoint(deadline, options.cancel_event)
            delay = retry_delay(attempt, None)
            error = None
            try:
                with self._http.stream(
                    method,
                    self._base + path,
                    params=parameters,
                    headers=headers,
                    content=content,
                    timeout=min(self.request_timeout, remaining),
                    follow_redirects=False,
                ) as response:
                    checkpoint(deadline, options.cancel_event)
                    if 300 <= response.status_code < 400:
                        raise APIError(response.status_code, "redirect_rejected")
                    chunks, size = [], 0
                    maximum = (
                        (self.max_export_bytes if binary else self.max_response_bytes)
                        if response.is_success
                        else 65536
                    )
                    for chunk in response.iter_bytes():
                        checkpoint(deadline, options.cancel_event)
                        size += len(chunk)
                        if size > maximum:
                            raise ClientError(
                                "response_too_large",
                                "API response exceeds the configured size limit.",
                            )
                        chunks.append(chunk)
                    raw = b"".join(chunks)
                    checkpoint(deadline, options.cancel_event)
                    if response.is_success:
                        if binary:
                            return raw
                        try:
                            result = json.loads(raw, parse_constant=reject_constant)
                            if not isinstance(result, dict):
                                raise ValueError()
                        except (ValueError, UnicodeError):
                            raise ClientError(
                                "invalid_response", "API returned an invalid JSON object."
                            ) from None
                        checkpoint(deadline, options.cancel_event)
                        return result
                    code, request_id = "http_error", response.headers.get("X-Request-Id")
                    message, details = None, None
                    try:
                        server_error = json.loads(raw).get("error", {})
                        candidate = server_error.get("code")
                        if isinstance(candidate, str) and re.fullmatch(
                            r"[a-z][a-z0-9_]{0,99}", candidate
                        ):
                            code = candidate
                        if isinstance(server_error.get("message"), str):
                            message = safe_error_value(server_error["message"], secrets)
                        details = safe_error_value(server_error.get("details"), secrets)
                    except (ValueError, AttributeError, UnicodeError):
                        pass
                    if any(value in code for value in secrets):
                        code = "http_error"
                    if (
                        not request_id
                        or not re.fullmatch(r"[A-Za-z0-9_-]{1,100}", request_id)
                        or any(value in request_id for value in secrets)
                    ):
                        request_id = None
                    error = APIError(
                        response.status_code, code, request_id, message=message, details=details
                    )
                    if response.status_code not in {429, 502, 503, 504}:
                        raise error
                    delay = retry_delay(attempt, response.headers.get("Retry-After"))
            except httpx.TransportError:
                checkpoint(deadline, options.cancel_event)
                error = ClientError(
                    "network_error",
                    "The API request failed; a submitted operation may have committed.",
                )
            if not retry or attempt == self.max_retries:
                raise error from None
            pause(delay, deadline, options.cancel_event)
        raise AssertionError("Unreachable retry state")


class Resource:
    def __init__(self, client: Transport):
        self._client = client

File: packages/typescript/src/transport.ts
export class ClientError extends Error {
  idempotency_key?: string;
  constructor(
    public readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = this.constructor.name;
  }
  toJSON() {
    return { code: this.code, message: this.message,
      ...(this.idempotency_key !== undefined ? { idempotency_key: this.idempotency_key } : {}) };
  }
}
export class APIError extends ClientError {
  constructor(
    public readonly status: number,
    code: string,
    public readonly requestId: string | null = null,
    message?: string,
    public readonly details: unknown = null,
  ) {
    super(code, message || `API request failed (HTTP ${status}, ${code}).`);
  }
  toJSON() {
    return {
      ...super.toJSON(),
      status: this.status,
      request_id: this.requestId,
      details: this.details,
    };
  }
}
export class RequestTimeout extends ClientError {
  constructor() {
    super(
      "request_timeout",
      "The request deadline expired; a submitted operation may have committed.",
    );
  }
}
export class WaitCancelled extends ClientError {
  constructor() {
    super(
      "wait_cancelled",
      "Local waiting was cancelled. The server evaluation was not cancelled.",
    );
  }
}
export type RequestOptions = { timeoutMs?: number; signal?: AbortSignal };
export type ClientOptions = {
  baseUrl?: string;
  apiKey?: string;
  workspaceId?: string;
  timeoutMs?: number;
  requestTimeoutMs?: number;
  maxRetries?: number;
  maxResponseBytes?: number;
  maxExportBytes?: number;
  allowHttp?: boolean;
  fetch?: typeof globalThis.fetch;
};
type Request = {
  body?: unknown;
  query?: object;
  retry: boolean;
  idempotency_key?: string;
  binary?: boolean;
  options?: RequestOptions;
};

export function positive(value: number, label: string): number {
  if (!Number.isFinite(value) || value <= 0 || value > 2_147_483_647)
    throw new ClientError(
      "invalid_configuration",
      `${label} must be a positive finite number no greater than 2147483647.`,
    );
  return value;
}
export function segment(value: string): string {
  if (
    typeof value !== "string" ||
    !/^[A-Za-z0-9_.:@+-]{1,200}$/.test(value) ||
    [".", ".."].includes(value)
  )
    throw new ClientError(
      "invalid_input",
      "Resource identifiers must be nonempty path-safe identifiers.",
    );
  return encodeURIComponent(value);
}
function normalizeBase(value: string, allowHttp: boolean): string {
  try {
    if (/[\u0000-\u0020\\]/.test(value)) throw new Error();
    const url = new URL(value);
    const local =
      url.hostname === "localhost" ||
      url.hostname === "[::1]" ||
      /^127\.\d+\.\d+\.\d+$/.test(url.hostname);
    if (
      url.username ||
      url.password ||
      url.search ||
      url.hash ||
      !["", "/v1"].includes(url.pathname.replace(/\/+$/, ""))
    )
      throw new Error();
    if (
      url.protocol !== "https:" &&
      !(url.protocol === "http:" && allowHttp && local)
    )
      throw new Error();
    return `${url.origin}/v1`;
  } catch {
    throw new ClientError(
      "invalid_configuration",
      "Set an HTTPS API origin or /v1 URL. Explicit HTTP is allowed only for loopback development.",
    );
  }
}
export function checkpoint(deadline: number, signal?: AbortSignal): number {
  if (signal?.aborted) throw new WaitCancelled();
  const remaining = deadline - performance.now();
  if (remaining <= 0) throw new RequestTimeout();
  return remaining;
}
export async function pause(
  delay: number,
  deadline: number,
  signal?: AbortSignal,
) {
  checkpoint(deadline, signal);
  const resumeAt = performance.now() + delay;
  // Timers can wake early. Recheck the monotonic clock so an early callback
  // cannot shorten Retry-After or turn a deadline-clipped wait into a retry.
  while (true) {
    const remaining = checkpoint(deadline, signal);
    const waiting = resumeAt - performance.now();
    if (waiting <= 0) return;
    await new Promise<void>((resolve, reject) => {
      const cleanup = () => {
        clearTimeout(timer);
        signal?.removeEventListener("abort", abort);
      };
      const abort = () => {
        cleanup();
        reject(new WaitCancelled());
      };
      const timer = setTimeout(
        () => {
          cleanup();
          resolve();
        },
        Math.ceil(Math.min(waiting, remaining)),
      );
      signal?.addEventListener("abort", abort, { once: true });
      if (signal?.aborted) abort();
    });
  }
}
function retryDelay(attempt: number, header: string | null): number {
  let delay = Math.min(5000, 250 * 2 ** attempt) + Math.random() * 250;
  if (header && header.length < 100) {
    const numeric = Number(header);
    const suggested = Number.isFinite(numeric)
      ? numeric * 1000
      : Date.parse(header) - Date.now();
    if (Number.isFinite(suggested)) delay = Math.max(delay, suggested);
  }
  return delay;
}
function secretValues(body: unknown): string[] {
  if (!body || typeof body !== "object") return [];
  return Object.entries(body).flatMap(([key, value]) => [
    ...(/key|secret|token|password/i.test(key) &&
    typeof value === "string" &&
    value
      ? [value]
      : []),
    ...secretValues(value),
  ]);
}
function safeErrorValue(value: unknown, secrets: string[], depth = 0): unknown {
  if (depth > 4) return "[truncated]";
  if (typeof value === "string") {
    let text = value;
    for (const secret of secrets) text = text.replaceAll(secret, "[redacted]");
    return text.replace(/[\u0000-\u001f\u007f-\u009f]/g, " ").slice(0, 2000);
  }
  if (Array.isArray(value))
    return value
      .slice(0, 100)
      .map((item) => safeErrorValue(item, secrets, depth + 1));
  if (value && typeof value === "object")
    return Object.fromEntries(
      Object.entries(value)
        .slice(0, 100)
        .map(([key, item]) => [
          (safeErrorValue(key, secrets) as string).slice(0, 100),
          /key|secret|token|password|authorization|cookie/i.test(key)
            ? "[redacted]"
            : safeErrorValue(item, secrets, depth + 1),
        ]),
    );
  return value === null ||
    typeof value === "number" ||
    typeof value === "boolean"
    ? value
    : null;
}
async function readBody(
  response: Response,
  maximum: number,
  deadline: number,
  signal?: AbortSignal,
): Promise<Uint8Array> {
  const reader = response.body?.getReader();
  if (!reader) return new Uint8Array();
  const chunks: Uint8Array[] = [];
  let size = 0;
  try {
    while (true) {
      checkpoint(deadline, signal);
      const { done, value } = await reader.read();
      if (done) break;
      size += value.byteLength;
      if (size > maximum)
        throw new ClientError(
          "response_too_large",
          "API response exceeds the configured size limit.",
        );
      chunks.push(value);
    }
    checkpoint(deadline, signal);
    return Buffer.concat(chunks);
  } finally {
    await reader.cancel().catch(() => undefined);
    reader.releaseLock();
  }
}

export class Transport {
  readonly timeoutMs: number;
  readonly requestTimeoutMs: number;
  readonly maxRetries: number;
  readonly maxResponseBytes: number;
  readonly maxExportBytes: number;
  #base: string;
  #key: string | undefined;
  #workspace: string | undefined;
  #fetch: typeof globalThis.fetch;
  constructor(options: ClientOptions = {}) {
    if (typeof window !== "undefined")
      throw new ClientError(
        "invalid_configuration",
        "Use workspace API keys only in trusted server applications.",
      );
    this.#base = normalizeBase(
      options.baseUrl ?? process.env.EVALROUTER_BASE_URL ?? "",
      options.allowHttp === true,
    );
    this.#key = options.apiKey ?? process.env.EVALROUTER_API_KEY;
    this.#workspace =
      options.workspaceId ?? process.env.EVALROUTER_WORKSPACE_ID;
    for (const [label, value] of [
      ["API key", this.#key],
      ["Workspace ID", this.#workspace],
    ]) {
      if (
        value !== undefined &&
        (typeof value !== "string" || !/^[\x21-\x7e]{1,4096}$/.test(value))
      )
        throw new ClientError(
          "invalid_configuration",
          `${label} must be printable non-space ASCII.`,
        );
    }
    this.timeoutMs = positive(options.timeoutMs ?? 120_000, "Timeout");
    this.requestTimeoutMs = positive(
      options.requestTimeoutMs ?? 30_000,
      "Request timeout",
    );
    this.maxRetries = options.maxRetries ?? 2;
    if (
      !Number.isInteger(this.maxRetries) ||
      this.maxRetries < 0 ||
      this.maxRetries > 5
    )
      throw new ClientError(
        "invalid_configuration",
        "maxRetries must be between 0 and 5.",
      );
    this.maxResponseBytes = options.maxResponseBytes ?? 64 * 1024 * 1024;
    this.maxExportBytes = options.maxExportBytes ?? 256 * 1024 * 1024;
    for (const value of [this.maxResponseBytes, this.maxExportBytes]) {
      if (!Number.isSafeInteger(value) || value < 1)
        throw new ClientError(
          "invalid_configuration",
          "Response size limits must be positive safe integers.",
        );
    }
    this.#fetch = options.fetch ?? globalThis.fetch;
  }
  async request<T>(method: string, path: string, request: Request): Promise<T> {
    const options = request.options ?? {};
    const deadline =
      performance.now() +
      positive(options.timeoutMs ?? this.timeoutMs, "Timeout");
    if (!path.startsWith("/") || path.startsWith("//"))
      throw new ClientError("invalid_input", "Invalid API resource path.");
    const url = new URL(this.#base + path);
    const headers: Record<string, string> = {
      Accept: request.binary ? "*/*" : "application/json",
      "User-Agent": "kimpton-evalrouter-typescript/0.1.0",
    };
    if (this.#key) headers.Authorization = `Bearer ${this.#key}`;
    if (this.#workspace) headers["X-Workspace-Id"] = this.#workspace;
    if (request.idempotency_key !== undefined) {
      if (
        typeof request.idempotency_key !== "string" ||
        !/^[\x20-\x7e]{1,128}$/.test(request.idempotency_key)
      )
        throw new ClientError(
          "invalid_input",
          "Idempotency keys must contain 1–128 printable ASCII characters.",
        );
      headers["Idempotency-Key"] = request.idempotency_key;
    }
    let body: string | undefined;
    try {
      body =
        request.body === undefined
          ? undefined
          : JSON.stringify(request.body, (_key, value: unknown) => {
              if (typeof value === "number" && !Number.isFinite(value))
                throw new Error();
              return value;
            });
      for (const [key, value] of Object.entries(request.query ?? {})) {
        if (value !== undefined && value !== null)
          url.searchParams.set(key, String(value));
      }
    } catch {
      throw new ClientError(
        "invalid_input",
        "Request data must be JSON-serializable.",
      );
    }
    if (body !== undefined) headers["Content-Type"] = "application/json";
    const secrets = [this.#key, ...secretValues(request.body)].filter(
      (v): v is string => !!v,
    );
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const remaining = checkpoint(deadline, options.signal);
      const controller = new AbortController();
      const abort = () => controller.abort();
      options.signal?.addEventListener("abort", abort, { once: true });
      const timer = setTimeout(
        abort,
        Math.ceil(Math.min(remaining, this.requestTimeoutMs)),
      );
      let error: ClientError;
      let delay = retryDelay(attempt, null);
      try {
        checkpoint(deadline, options.signal);
        const response = await this.#fetch(url, {
          method,
          headers,
          body,
          signal: controller.signal,
          redirect: "manual",
          credentials: "omit",
        });
        if (response.status >= 300 && response.status < 400) {
          await response.body?.cancel().catch(() => undefined);
          throw new APIError(response.status, "redirect_rejected");
        }
        const raw = await readBody(
          response,
          response.ok
            ? request.binary
              ? this.maxExportBytes
              : this.maxResponseBytes
            : 65536,
          deadline,
          options.signal,
        );
        if (response.ok) {
          if (request.binary) return raw as T;
          try {
            const value: unknown = JSON.parse(
              new TextDecoder("utf-8", { fatal: true }).decode(raw),
            );
            if (!value || typeof value !== "object" || Array.isArray(value))
              throw new Error();
            checkpoint(deadline, options.signal);
            return value as T;
          } catch (e) {
            if (e instanceof ClientError) throw e;
            throw new ClientError(
              "invalid_response",
              "API returned an invalid JSON object.",
            );
          }
        }
        let code = "http_error",
          requestId = response.headers.get("X-Request-Id");
        let message: string | undefined,
          details: unknown = null;
        try {
          const serverError = JSON.parse(new TextDecoder().decode(raw))?.error;
          const candidate: unknown = serverError?.code;
          if (
            typeof candidate === "string" &&
            /^[a-z][a-z0-9_]{0,99}$/.test(candidate)
          )
            code = candidate;
          if (typeof serverError?.message === "string")
            message = safeErrorValue(serverError.message, secrets) as string;
          details = safeErrorValue(serverError?.details, secrets);
        } catch {
          /* Untrusted response text does not enter exceptions. */
        }
        if (secrets.some((secret) => code.includes(secret)))
          code = "http_error";
        if (
          !requestId ||
          !/^[A-Za-z0-9_-]{1,100}$/.test(requestId) ||
          secrets.some((secret) => requestId!.includes(secret))
        )
          requestId = null;
        error = new APIError(
          response.status,
          code,
          requestId,
          message,
          details,
        );
        if (![429, 502, 503, 504].includes(response.status)) throw error;
        delay = retryDelay(attempt, response.headers.get("Retry-After"));
      } catch (e) {
        checkpoint(deadline, options.signal);
        if (e instanceof ClientError) throw e;
        error = new ClientError(
          "network_error",
          "The API request failed; a submitted operation may have committed.",
        );
      } finally {
        clearTimeout(timer);
        options.signal?.removeEventListener("abort", abort);
      }
      if (!request.retry || attempt === this.maxRetries) throw error;
      await pause(delay, deadline, options.signal);
    }
    throw new Error("Unreachable retry state");
  }
}

export class Resource {
  constructor(protected readonly client: Transport) {}
}
